File: akira-web/backend/tools/file_management/list_dir.py
"""
List directory contents with metadata (name, type, size). Helps discover structure
without blind reads. Paths are canonicalized and restricted to the project workspace.
"""
import logging
from pathlib import Path

from backend.core.file_access import resolve_path

logger = logging.getLogger(__name__)

TOOL_DEF = {
    "name": "list_dir",
    "description": "List contents of a directory with name, type (file/dir), and size (files only). Use relative paths from the project root. Paths are canonicalized and must stay inside the project. Use to discover structure before read_file.",
    "input_schema": {
        "type": "object",
        "properties": {
            "dir_path": {
                "type": "string",
                "description": "Path to the directory (relative to project root or absolute within project). Use '.' or omit for project root.",
            },
            "max_entries": {
                "type": "integer",
                "description": "Maximum number of entries to return (default: 200). Use to limit output for large directories.",
            },
        },
        "required": [],
    },
    "default_enabled": True,
}


def call_tool(tool_input: dict, context=None):
    dir_path = (tool_input.get("dir_path") or ".").strip()
    max_entries = tool_input.get("max_entries")
    if max_entries is not None:
        try:
            max_entries = max(1, int(max_entries))
        except (TypeError, ValueError):
            max_entries = 200
    else:
        max_entries = 200

    resolved = resolve_path(dir_path)
    if resolved is None:
        logger.warning("Rejected path outside workspace: %s", dir_path)
        return 200, {
            "success": False,
            "error": "Path is outside the project workspace.",
            "path": dir_path,
        }

    path_str = str(resolved)
    if not resolved.exists():
        return 200, {"success": False, "error": "Path not found.", "path": path_str}
    if not resolved.is_dir():
        return 200, {"success": False, "error": "Not a directory.", "path": path_str}

    try:
        entries = []
        for p in sorted(resolved.iterdir()):
            if len(entries) >= max_entries:
                entries.append({
                    "name": "...",
                    "type": "truncated",
                    "size": None,
                    "note": f"Limited to {max_entries} entries.",
                })
                break
            try:
                stat = p.stat()
                entry = {
                    "name": p.name,
                    "type": "dir" if p.is_dir() else "file",
                    "size": stat.st_size if p.is_file() else None,
                }
                entries.append(entry)
            except OSError:
                entries.append({"name": p.name, "type": "unknown", "size": None})

        logger.info("Listed directory: %s (%s entries)", path_str, len(entries))
        return 200, {
            "success": True,
            "path": path_str,
            "entries": entries,
        }
    except Exception as e:
        logger.error("Error listing %s: %s", path_str, e, exc_info=True)
        return 200, {"success": False, "error": str(e), "path": path_str}

File: akira-desktop/electron/tools/file-operations/list-dir.js
/**
 * List Directory Tool
 * list_dir - List contents of a directory
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// Workspace root - default to user's home directory
let WORKSPACE_ROOT = os.homedir();

function setWorkspaceRoot(root) {
  WORKSPACE_ROOT = root;
}

/**
 * Resolve and validate path stays within workspace
 */
function resolvePath(filePath) {
  if (!filePath) return null;

  // Handle absolute paths
  let resolved;
  if (path.isAbsolute(filePath)) {
    resolved = path.normalize(filePath);
  } else {
    resolved = path.normalize(path.join(WORKSPACE_ROOT, filePath));
  }

  return resolved;
}

const definitions = [
  {
    name: 'list_dir',
    description: 'List contents of a directory with name, type (file/dir), and size.',
    input_schema: {
      type: 'object',
      properties: {
        dir_path: {
          type: 'string',
          description: 'Path to the directory. Use "." or omit for current workspace.',
        },
        max_entries: {
          type: 'integer',
          description: 'Maximum entries to return (default: 200)',
        },
      },
      required: [],
    },
  },
];

const handlers = {
  async list_dir(input) {
    const dirPath = resolvePath(input.dir_path || '.');
    if (!dirPath) {
      return { success: false, error: 'Invalid directory path' };
    }

    const maxEntries = Math.min(input.max_entries || 200, 500);

    if (!fs.existsSync(dirPath)) {
      return { success: false, error: 'Path not found', path: dirPath };
    }

    const stats = fs.statSync(dirPath);
    if (!stats.isDirectory()) {
      return { success: false, error: 'Not a directory', path: dirPath };
    }

    try {
      const items = fs.readdirSync(dirPath);
      const entries = [];

      for (const name of items.slice(0, maxEntries)) {
        try {
          const itemPath = path.join(dirPath, name);
          const itemStats = fs.statSync(itemPath);
          entries.push({
            name,
            type: itemStats.isDirectory() ? 'dir' : 'file',
            size: itemStats.isFile() ? itemStats.size : null,
          });
        } catch {
          entries.push({ name, type: 'unknown', size: null });
        }
      }

      if (items.length > maxEntries) {
        entries.push({
          name: '...',
          type: 'truncated',
          note: `Limited to ${maxEntries} entries. Total: ${items.length}`,
        });
      }

      return { success: true, path: dirPath, entries };
    } catch (error) {
      return { success: false, error: error.message, path: dirPath };
    }
  },
};

module.exports = { definitions, handlers, setWorkspaceRoot };
